README.md
# StTools

Библиотека StTools предназначена для поддержки CLI- и Rails-приложений. Она обеспечивает
различное преобразование строк с учетом русского языка, формирование человеко-удобного представления времени и чисел,
получения разных системных характеристик CLI-приложений (память, размер экрана терминала),
управление прогресс-баром CLI-приложений.

## Установка

Добавить в Gemfile:

```ruby
gem 'st_tools'
```

Установить гем cредствами Bundler:

    $ bundle

Или установить его отдельно:

    $ gem install st_tools

# Зависимости

Для работы гема требуется Ruby не младше версии 2.2.0. StTools не привязана к Rails и может использоваться в CLI-приложениях.
 Класс StTools::ProgressBar является надстройкой над гемом 'progressbar-ruby'.

## Использование

Часть методов библиотеки поддерживают русскую и английскую локализацию.
Поскольку все методы StTools могут вызываться без создания экземпляра класса - вызов конструктура класса отсутствует.
Ряд методов библиотеки требуют настройки локализации. Для этого, в приложении Ruby CLI нужно использовать:

```ruby
StTools.configure { |config| config.locale = :ru }
```

В приложении Rails необходимо создать файл `/initializers/st_tools.rb`, и включить в него:
```ruby
StTools.configure do |config|
  config.locale = :ru
end
```

### StTools::Human

Вы вызываете в любой момент StTools::Human.memory и узнаете текущий размер памяти, занимаемый приложением (процессом)

```ruby
StTools.configure { |config| config.locale = :ru }
StTools::Human.memory               # => 14 кбайт
StTools::Human.memory               # => 45,3 Мбайт
StTools::Human.memory               # => 2,6 Гбайт
```

Вы вызываете StTools::Human.bytes и перевести любое значение в человеко-удобный вид. Метод можно использовать для показа размера файла

```ruby
StTools.configure { |config| config.locale = :ru }
StTools::Human.bytes(345)           # => 345 байт
StTools::Human.bytes(14653)         # => 14,5 кбайт
StTools::Human.bytes(23653763)      # => 23,4 Мбайт
```

Вы вызываете StTools::Human.number для перевода любого значения в человеко-удобный вид.
Метод можно использовать для показа, например, суммы денег

```ruby
StTools.configure { |config| config.locale = :ru }
StTools::Human.number(345)           # => 345
StTools::Human.number(14653)         # => 14,6 тыс.
StTools::Human.number(23653763)      # => 23,7 млн.
```

Вы вызываете StTools::Human.pretty_number для красивого отображения числа с пробелами. При этом можно указать
число цифр после запятой, и автоматически конвертировать точку в звпятую

```ruby
StTools.configure { |config| config.locale = :ru }
StTools::Human.pretty_number(345)                       # => 345
StTools::Human.pretty_number(345, round: 2)             # => 345,00
StTools::Human.pretty_number(75345, round: 1)           # => 75 345,0
StTools::Human.pretty_number(nil)                       # => 0
StTools::Human.pretty_number('1675345.763', round: 1)   # => 1 675 345,7
```

Вы имеете возможность узнать разницу между текущим временем и временем какого-либо события.

```ruby
StTools.configure { |config| config.locale = :ru }
StTools::Human.human_ago(DateTime.new(2014,12,31), true)        # => 4 месяца 21 день назад
StTools::Human.human_ago(DateTime.new(2013,08,01), false)       # => 1 год 8 месяца
StTools::Human.human_ago(Time.now - 15, true)                   # => 15 секунд назад
```

```ruby
StTools.configure { |config| config.locale = :en }
StTools::Human.human_ago(DateTime.new(2014,12,31), true)        # => 4 months 21 days ago
StTools::Human.human_ago(DateTime.new(2013,08,01), false)       # => 1 year 8 months
StTools::Human.human_ago(Time.now - 15, true)                   # => 15 seconds ago
```

Устанавливая флаг ```ago=true```, метод добавит слово "назад" или "ago" в конце фразы.

Результирующая фраза всегда состоит из одного показатели (при разнице времени событий меньше минуты)
или из двух (минуты/секунды, часы/минуты, дни/часы, месяцы/дни, года/месяцы)

Иногда возникает другой кейс, когда нужно рассчитать разницу во времени не относительно текущей даты,
а относительно любых двух произвольных событий. Для этого может использоваться метод, в который передается
не временная метка, а просто количество секунд между двумя событиями.

```ruby
StTools.configure { |config| config.locale = :ru }
StTools::Human.seconds_ago(DateTime.new(2014,12,31), true)        # => 4 months 21 days ago
StTools::Human.seconds_ago(DateTime.new(2013,08,01), false)       # => 1 year 8 months
StTools::Human.seconds_ago(Time.now - 15, true)                   # => 15 seconds ago
```


Вы можете форматировать дату и время в соответствии с правилами русского и английского языка.

```ruby
StTools.configure { |config| config.locale = :ru }
StTools::Human.format_time(Time.now, :full, :full)              # => 28 апреля 2015 г. 10:45:23
StTools::Human.format_time(Time.now, :date, :full)              # => 28 апреля 2015 г.
StTools::Human.format_time(Time.now, :time, :full)              # => 10:45:23
StTools::Human.format_time(Time.now, :full, :short)             # => 28/04/2015 10:45
StTools::Human.format_time(Time.now, :date, :short)             # => 28/04/2015
StTools::Human.format_time(Time.now, :time, :short)             # => 10:45
```

```ruby
StTools.configure { |config| config.locale = :en }
StTools::Human.format_time(Time.now, :full, :full)              # => April 28, 2015 09:45:23 am
StTools::Human.format_time(Time.now, :date, :full)              # => April 28, 2015
StTools::Human.format_time(Time.now, :time, :full)              # => 09:45:23 am
StTools::Human.format_time(Time.now, :full, :short)             # => 04/28/2015 09:45 am
StTools::Human.format_time(Time.now, :date, :short)             # => 04/28/2015
StTools::Human.format_time(Time.now, :time, :short)             # => 09:45 am
```

Новая версия данной функции имеет более прстой и понятный синтаксис. Первый параметр - формат даты, второй - времени.

```ruby
StTools.configure { |config| config.locale = :ru }
StTools::Human.format_time2(Time.now, :human, :full)               #=> "30 апреля 2015 г. 08:54:34"
StTools::Human.format_time2(Time.now, :human, :full, god: false)   #=> "30 апреля 2015 08:54:34"
StTools::Human.format_time2(Time.now, :human, :short)              #=> "30 апреля 2015 г. 8:54"
StTools::Human.format_time2(Time.now, :human, :none)               #=> "30 апреля 2015 г."
StTools::Human.format_time2(Time.now, :full, :full)                #=> "30/04/2015 08:54:34"
StTools::Human.format_time2(Time.now, :short, :short)              #=> "30/04/15 8:54"
StTools::Human.format_time2(Time.now, :none, :full)                #=> "08:54:34"
StTools::Human.format_time2(Time.now, :none, :short)               #=> "8:54"
````

Вы можете подмешать модуль `StTools::Module` в классы String, Integer, Time.

```ruby
StTools::Setup.setup(:ru)

class String
   include StTools::Module::String
end

(Time.now - 15)::Human_ago                         # => 15 секунд назад
Time.now.format_date                              # => 28 апреля 2015 г.
Time.now.format_date(:short)                      # => 28/04/2015
```

### StTools::String

#### translit

Транслитерация строки из русского на английский

```ruby
StTools::String.translit('Жмеринка')                  # => Zhmerinka
```

#### delat

Иногда пользователи вводят английские буквы, похожие на русские, чаще все - букву 'C'. Метод delat заменяет такие символы на русские

```ruby
StTools::String.delat('Соль')                         # => Соль
```

#### upcase/downcase

Перевести строку в верхний или нижний регистр. Используется метод, не использующий :mb_chars

```ruby
StTools::String.downcase('Москва')                    # => москва
StTools::String.upcase('Москва')                      # => МОСКВА
```

#### normalize

Иногда надо нормализовать строку и привести ее к виду, удобному для машинной обработки.
Метод normalize последовательно делает: `strip`, `delat`, `downcase`. Кроме того, метод корректно заменит
на обычные пробелы "короткие" пробелы, получающиеся после преобразования тега HTML `&nbsp;` в пробел. 

```ruby
StTools::String.normalize('   Москва  ')              # => "москва"
```

Внимание!!! Функцию следует использовать с осторожностью, и не забывать, что для ряда задач она не подходит. 
В частности, выполнение `StTools::String.normalize('TRUE')` даст значение `true`, в котором буква `е` будет
русская (применена функция `delat`). 

#### caps

Следующий метод приводит строку к нижнему регистру, но делает первую букву заглавной.
Удобно для нормализации имени и фамилии, вводимых пользователем

```ruby
StTools::String.caps('санКТ-петерБург')              # => "Санкт-Петербург"
StTools::String.caps('1-Я владимирская')             # => "1-я Владимирская"
StTools::String.caps('stan.zhuravlev@mail.ru')       # => "Stan.Zhuravlev@mail.Ru"
StTools::String.caps('а.с. ПУШКИН')                  # => "А.С. Пушкин"
StTools::String.caps('ivan_petrov')                  # => "Ivan_Petrov"
StTools::String.caps('н.а.римский-корсаков')         # => "Н.А.Римский-Корсаков"

StTools::String.caps('н.а.римский-корсаков пришел в гости к в.и. ленину на улицу героев 1812-го года')
  #=> "Н.А.Римский-Корсаков Пришел В Гости К В.И. Ленину На Улицу Героев 1812-го Года"
```

#### hide

Для реализации деморежимов различных приложений иногда надо скрыть реальный результат. Для это можно
использовать метод hide

```ruby
StTools::String.hide('Краснодар')              # => "Кра*но*ар"
```

#### to_bool

В Ruby отсутствует метод перевода строки пользователя в тип Boolean.

```ruby
StTools::String.to_bool("True")    # => true
StTools::String.to_bool("trUE")    # => true
StTools::String.to_bool("on")      # => true
StTools::String.to_bool("1")       # => true
StTools::String.to_bool("Да")      # => true
StTools::String.to_bool("Yes")     # => true
StTools::String.to_bool("false")   # => false
StTools::String.to_bool("fALse")   # => false
StTools::String.to_bool("oFF")     # => false
StTools::String.to_bool("0")       # => false
```

Когда в Rails-приложение передается массив params, то значение bool обычно выглдяит строкой. При этом часть значений
опциональные, то есть пользователь может не ввести его. Для этого можно использовать второй параметр функции, по
умолчанию имеющий значение false.

```
params = { opt1: true }
StTools::String.to_bool(params[:opt1], false)      #=> true
StTools::String.to_bool(params[:opt2], true)       #=> true
```

#### split

Если пользователь введет в качестве входного параметра перечисление, следующий метод разобъет его на токены,
с нормализацией при необходимости

```ruby
StTools::String.split('ТАНЯ, маша, петя', ',')                                 # => ['ТАНЯ', 'маша', 'петя']
StTools::String.split('ТАНЯ, маша, петя', ',', normalize: true)                # => ['таня', 'маша', 'петя']
StTools::String.split('ТАНЯ, маша, петя', ',', normalize: true, sort: true)    # => ['маша', 'петя', 'таня']
```

#### to_range

Если от пользователя требуется ввести численные значения в виде перечисления или диапазоне, нужно использовать
метод to_range

```ruby
StTools::String.to_range('54, 3-6, 5, 1')                                 # => [54, 3, 5, 6, 5, 1]
StTools::String.to_range('54, 3-6, 5, 1', uniq: true)                     # => [54, 3, 5, 6, 1]
StTools::String.to_range('54, 3-6, 5, 1', uniq: true, sort: true)         # => [1, 3, 5, 6, 54]
```

#### pretty_list

Следующий метод `pretty_list` позволяет сделать human-подобное оформление списков-перечислений.

```ruby
StTools::Setup.setup(:ru)
'Мы работаем ' + StTools::String.pretty_list([4,5,6]) + ' мая'
    # => "Мы работаем 4, 5 и 6 мая"
'Доехать можно на трамваях ' + StTools::String.pretty_list([67, 89], union: :or) + ' маршрутов'
    # => "Доехать можно на трамвае 67 или 89 маршрутов"
'Скидка составит ' + StTools::String.pretty_list([100, 200], union: :or, pretag: '<em>', afttag: '</em>') + ' руб.'
    # => "Скидка составит <em>100</em> или <em>200</em> руб."

StTools::Setup.setup(:en)
'Discount is ' + StTools::String.pretty_list([100, 200], union: :or, pretag: '<em>$', afttag: '</em>')
    # => "Discount is <em>$100</em> or <em>$200</em>"
```

#### prune

В ряде случаев необходимо обрезать строку таким образом, чтобы на экране уместилась часть строки, и многоточие,
которе показывает, что строка была сокращена. Для этого можно использовать метод `StTools::String.prune`.

```ruby
StTools::String.prune("1234567890", 20)                   #=> "1234567890"
StTools::String.prune("1234567890", 8)                    #=> "12345..."
StTools::String.prune("1234567890", 8, endwidth: '---')   #=> "12345---"
StTools::String.prune("Привет мир!", 12)                  #=> "Привет ми..."
StTools::String.prune("Привет мир!", 12, words: true)     #=> "Привет..."
```

Следует обратить внимание на два момента. Во-первых, строка обрезается до указанной длины ПЛЮС длина завершающего многточия.
То есть при указании `StTools::String.prune("1234567890", 5)` результат будет `"12..."`. Во-вторых, если указать параметр
`word: true`, то произойдет более "умное" обрезание фразы - не посреди слова (как показано в примере выше), а по границе
слова, то есть по пробелу между словами в меньшую сторону.

#### to_float

Метод `to_float` позволяет переводить различные строки, введенные пользователем, в тип `Float` языка Ruby.

```ruby
StTools::String.to_float('123.45678')                                 #=> 123.45678
StTools::String.to_float('123.474565', round: 2)                      #=> 123.47
StTools::String.to_float('123,474565', round: 2)                      #=> 123.47
StTools::String.to_float('   123,47456564', round: 2)                 #=> 123.47
StTools::String.to_float('   10 123,47456', round: 2)                 #=> 10123.47
StTools::String.to_float(' -  10 123,474565', round: 2)               #=> -10123.47
```

Если у Вас уже есть число в формате `Float` или `Integer`, то все равно можно вызвать метод `to_float`. При этом
произойдет только округление числа в соответствии с параметром `round`.

```ruby
StTools::String.to_float(145.5667, round: 2)                          #=> 145.57
StTools::String.to_float(23, round: 2)                                #=> 23
```

Возможны ситуации, когда параметр переданный является `nil` или строкой, не содержащей числа. В этом случае возможны
две модели поведения. Первая (установленная по умолчанию), выбрасывает Exception в случае, если переданная на вход
метода строка не содержит цифр. Другая модель предусматривает возвращаение нулевого значения при ошибке. Для этого
нужно установить параметр `stop` в значение `false`.

```ruby
StTools::String.to_float(nil) rescue 'fail')                          #=> "fail"
StTools::String.to_float(nil, stop: false) rescue 'fail')             #=> 0
StTools::String.to_float('no digits') rescue 'fail')                  #=> "fail"
StTools::String.to_float('no digits', stop: false) rescue 'fail')     #=> 0
```

#### Расширение класса String (include StTools)

Вы можете подмешать модуль `StTools::Module` в классы String, Integer, Time.

```ruby
StTools::Setup.setup(:ru)

class String
   include StTools::Module::String
end

'санКТ-петерБург'.caps                         # => Санкт-Петербург
'санКТ-петерБург'.downcase                     # => санкт-петербург
'санКТ-петерБург'.upcase                       # => САНКТ-ПЕТЕРБУРГ
'on'.to_bool                                   # => true
'34, 2-5, 8, 1'.to_range(sort: true)           # => [1, 2, 4, 5, 8, 34]
'Краснодар'.hide                               # => Кра*но*ар
```

### StTools::System

Узнать объем оперативной памяти, занятый приложением

```ruby
StTools::System.memory                                 # => 145734
```

Узнать ширину и высоту текущего окна терминала в CLI-режиме

```ruby
StTools::System.screen(:width)               # => 114
StTools::System.screen(:height)              # => 25
```

### StTools::ProgressBar

Данная бибилиотека базируется на геме 'progressbar-ruby' (https://github.com/jfelchner/ruby-progressbar), но надстройка над ним
упрощает использование библиотеки, и исключает некоторые проблемы.

Библиотека содержит две функции. Первая - инициирует начало отображения прогресс-бара

```ruby
@bar = ::StTools::ProgressBar.new(title: "Тестовый запуск",
                                  footer: "Занято [memory], выполнено за [executed_at] сек.",
                                  max: 10000000)
```

Задается заголовок, который будет отображен перед прогресс-баром. Затем указывается футер, который отобразиться как
только прогресс станет 100%. Футер содержит две шаблона:
- [memory], который будет заменен на результат вызова StTools::Human.memory
- [executed_at], который будет заменен на время в секундах завершенного процесса

Значение :max содержит финальное значение прогресса. Дополнительно рекомендуется задать значение :step, которое
показывает, на каком шаге в реальности должен меняться процент выполнения операции.

Затем, в цикле, который обеспечивает выполнение длительной операции, необходимо менять значение progress.

```ruby
10000010.times do |val|
  @bar.progress = val
end
```


## Contributing

1. Fork it ( https://github.com/[my-github-username]/st_tools/fork )
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create a new Pull Request


CHANGELOG.md
## 0.4.3 (2019-01-17)

* Исправлена ошибка рекурсии при использовании Ruby 2.6.0

## 0.4.2 (2017-01-04)

* Изменено сообщение DEPRECATED для метода setup

## 0.4.1 (2016-09-10)

* В функции `StTools::Human.format_time2` исправлена ошибка с отображением " г."

## 0.4.0 (2016-09-09)

* В метод `StTools::String.normalize` добавлен параметр `delat`. Изменение потребовалось после того, как при использовании функции normalize для генерации `slug` имен в одном из проектов выяснилось, что в URL часть английских букв заменены на русские.
* Вместо метода  `StTools::Setup.setup` применен `StTools.configure { |config| config.locale = :ru }`. Подробнее - см. README.md
* Написана функция `StTools::Human.format_time2` с более простыми и понятными ключами запуска
* Методы `StTools::Human.bytes` и `StTools::Human.numbers` локализованы через I18n

## 0.3.18 (2016-05-12)

* Исправлена ошибка привязки к Ruby 2.2.x

## 0.3.17 (2016-04-23)

* Добавлены методы `StTools::Fias.other?` и `StTools::Fias.other_aoguid`
* Незначительные улучшения

## 0.3.16 (2015-12-10)

* Добавлен метод `StTools::Human.pretty_number` для красивого форматирования числа

## 0.3.15 (2015-12-05)

* Незначительные улучшения

## 0.3.14 (2015-12-03)

* Скорректировано поведение прогресс-бара при приближении значения к максимальному уровню
* Добавлена библиотека стран для совместимости с post-API.ru

## 0.3.13 (2015-11-14)

* Исправлена ошибка с расчетом размера задействованной памяти (вместо килобайт, теперь возвращаются байты)

## 0.3.12 (2015-10-28)

* Были перепутаны местами aoguid Севастополя и Байконура

## 0.3.11 (2015-07-09)

* Функция `StTools::String.normalize` теперь удаляет и лишние "короткие" пробелы (&nbsp;)
* Добавлена функция `StTools::Fias.distance`, рассчитывающая расстояние между двумя географическиими точками

## 0.3.10 (2015-06-06)

* Исправлена небольшая ошибка в `StTools::String.prune` (добавил "жадность" в регулярку)
* Добавлена функция `StTools::String.to_float` для преобразования строк в числа с плавающей запятой

## 0.3.9 (2015-06-04)

* Исправлена потенциальная уязвимость регулярных выражений - `/^$/` заменены на `/\A\z/`.
* Добавлена функция `StTools::String.prune` для обрезания строки с добавлением многоточья

## 0.3.8 (2015-05-31)

* Добавлен метод `StTools::String.pretty_list`, оформляющий перечисление, например "Паша, Миша и Саша"

## 0.3.7 (2015-05-21)

* В метод `StTools::String.to_bool` добавлена возможность указывать значение по умолчанию для значений `nil`.

## 0.3.6 (2015-05-21)

* Доработан метод `StTools::String.caps`, добавлена более интеллектуальная обработка входных строк
* Исправлена ошибка показа футера прогресс бара, если executed_at > 60 секунд
* Исправлена ошибка метода `StTools::System.exename` при вызове скрипта с параметрами командной строки

## 0.3.4 (2015-05-01)

* Добавлено удаление функций :downcase, :upcase стандартного модуля String при подмешивании (include) StTools
* Название метода .words_in_ago изменено на .human_ago
* Добавлен метод .seconds_ago, форматирующий строку для количества секунд между событиями

## 0.3.3 (2015-04-30)

* Название метода .hide_text изменено на .hide
* Доработана документация
* Исправлены результаты работы функций .split и .to_range при передаче в них nil
* В метод .to_bool добавлены значения "да" и "yes" в качестве значений, являющихся true


lib/st_tools/human.rb
module StTools
  class Human


    # Метод форматирует число, добавляя суффиксы 'тыс.', 'млн.' и пр.
    # Предварительно необходимо вызвать StTools.configure { |config| config.locale = :ru }.
    #
    # @param [Integer] val исходное числовое значение
    # @return [String] строка вида "512,4 тыс."
    # @example Примеры использования
    #   StTools.configure { |config| config.locale = :ru }
    #   StTools::Human.number(123)           #=> "123"
    #   StTools::Human.number(14563)         #=> "14 тыс."
    #   StTools::Human.number(763552638)     #=> "763.6 млн."
    def self.number(val)
      StTools.configuration.numbers_array.each_pair do |e, s|
        if val < s
          if s <= 1000*1000
            return "#{(val.to_f / (s / 1000)).round(0)}#{I18n.t('st_tools.numbers_separator', locale: StTools.configuration.locale)}#{e}"
          else
            return "#{(val.to_f / (s / 1000)).round(1)}#{I18n.t('st_tools.numbers_separator', locale: StTools.configuration.locale)}#{e}"
          end
        end
      end
    end

    # Метод форматирует число, добавляя суффиксы 'кбайт', 'Мбайт' и др.
    # Предварительно необходимо вызвать StTools.configure { |config| config.locale = :ru }.
    #
    # @param [Integer] val исходное числовое значение в байтах
    # @return [String] строка вида "512,4 Мбайт"
    # @example Примеры использования
    #   StTools.configure { |config| config.locale = :ru }
    #   StTools::Human.bytes(123)           #=> "123 байта"
    #   StTools::Human.bytes(14563)         #=> "14 кбайт"
    #   StTools::Human.bytes(763552638)     #=> "728.2 Мбайт"
    def self.bytes(val)
      # noinspection RubyStringKeysInHashInspection
      StTools.configuration.bytes_array.each_pair do |e, s|
        if val < s
          if s <= 1024*1024
            return "#{(val.to_f / (s / 1024)).round(0)} #{e}"
          else
            return "#{(val.to_f / (s / 1024)).round(1)} #{e}"
          end
        end
      end
    end

    # Метод возвращает форматированную строку с объемом памяти, занимаемым текущим процессом (pid).
    #
    # @return [String] строка вида "512,4 Мбайт"
    def self.memory
      val = ::StTools::System.memory
      return self.bytes(val)
    end

    # Метод переводит DateTime в строку на русском или иных языках вида "4 дня 23 часа назад".
    # Предварительно необходимо вызвать StTools.configure { |config| config.locale = :ru }.
    #
    # @param [DateTime] time время и дата
    # @param [Boolean] ago true, если надо добавить слово "назад" в конец строки
    # @return [String] строка вида "3 дня 12 часов" или "3 дня 12 часов назад"
    # @example Примеры использования
    #   StTools.configure { |config| config.locale = :ru }
    #   StTools::Human.human_ago(Time.now - 23, true)       #=> "23 секунды назад"
    #   StTools::Human.human_ago(Time.now - 24553, false)   #=> 6 часов 49 минут"
    #   StTools::Human.human_ago(Time.now)                  #=> "сейчас"
    def self.human_ago(time, ago = true)
      now = self.to_time(Time.now.strftime('%Y-%m-%d %H:%M:%S UTC'))
      slf = self.to_time(time.strftime('%Y-%m-%d %H:%M:%S UTC'))
      secs = (now - slf).to_i
      return I18n.t('common.ago.very_long', locale: StTools.configuration.locale) if time.year < 1800
      return I18n.t('common.ago.just_now', locale: StTools.configuration.locale) if secs > -1 && secs < 1
      return '' if secs <= -1
      pair = self.ago_in_words_pair(secs)
      pair << I18n.t("common.ago.ago_word", locale: StTools.configuration.locale) if ago == true
      pair.join(' ')
    end

    # Метод принимает параметр - количество секунд между двумя любыми событиями в секундах,
    # и переводит их в строку на русском или иных языках вида "4 дня 23 часа назад".
    # Предварительно необходимо вызвать StTools.configure { |config| config.locale = :ru (или :en) }.
    #
    # @param [DateTime] sesc количество секунд
    # @param [Boolean] ago true, если надо добавить слово "назад" в конец строки
    # @return [String] строка вида "3 дня 12 часов" или "3 дня 12 часов назад"
    # @example Примеры использования
    #   StTools.configure { |config| config.locale = :ru }
    #   StTools::Human.seconds_ago(23, true)       #=> "23 секунды назад"
    #   StTools::Human.seconds_ago(24553, false)   #=> 6 часов 49 минут"
    #   StTools::Human.seconds_ago(0)              #=> "сейчас"
    def self.seconds_ago(secs, ago = true)
      secs_i = secs.to_i
      return I18n.t('common.ago.just_now', locale: StTools.configuration.locale) if secs_i > -1 && secs_i < 1
      return '' if secs_i <= -1
      pair = self.ago_in_words_pair(secs_i)
      pair << I18n.t("common.ago.ago_word", locale: StTools.configuration.locale) if ago == true
      pair.join(' ')
    end

    # Метод переводит DateTime в строку на русском или иных языках. Предварительно необходимо вызвать
    # StTools.configure { |config| config.locale = :ru (или :en) }.
    #
    # DEPRECATED
    #
    # @param [DateTime] time исходные время и дата
    # @param [Sym] what формат возвращаемого результата, принимает одно из следующих значений
    # @option :full форматирует дату и время (по умолчанию)
    # @option :date форматирует только дату
    # @option :time форматирует только время
    # @param [Sym] type форма в которой возращать результат: длинная ("28 апреля 2015 г. 10:34:52") или короткая ("28/04/2015 10:34")
    # @option :full длинна форма
    # @option :short короткая форма
    # @return [String] строка с форматированными датой и временем
    # @example Примеры использования
    #   StTools.configure { |config| config.locale = :ru }
    #   StTools::Human.format_time(Time.now, :full, :full)       #=> "30 апреля 2015 г. 08:54:34"
    #   StTools::Human.format_time(Time.now, :date, :full)       #=> "30 апреля 2015 г."
    #   StTools::Human.format_time(Time.now, :time, :full)       #=> "08:54:34"
    #   StTools::Human.format_time(Time.now, :full, :short)      #=> "30/04/2015 08:55"
    #   StTools::Human.format_time(Time.now, :date, :short)      #=> "30/04/2015"
    #   StTools::Human.format_time(Time.now, :time, :short)      #=> "08:55"
    def self.format_time(time, what, type)
      unless [:full, :date, :time].include?(what)
        warn "WARNING: what ':#{what.to_s}' must be in [:full, :date, :time]. Use ':full' now (at line #{__LINE__} of StTools::#{File.basename(__FILE__)})"
        what = :full
      end
      return I18n.l(time, :format => "#{what.to_s}_#{type.to_s}".to_sym, locale: StTools.configuration.locale)
    end

    # Метод переводит DateTime в строку на русском или иных языках. Предварительно необходимо вызвать
    # StTools.configure { |config| config.locale = :ru (или :en) }.
    #
    # @param [DateTime] timestamp исходные время и дата
    # @param [Sym] date формат возвращаемого результата, принимает одно из следующих значений
    # @option :human форматирует дату и время в формате "21 марта 2011"
    # @option :full форматирует дату и время в формате "21.03.2016"
    # @option :short форматирует только дату в формате "21.3.16"
    # @option :none игнорирует значение даты
    # @param [Sym] time форма в которой возращать результат
    # @option :full длинна форма в формате "08:23:03"
    # @option :short короткая форма в формате "8:23", "22:34", "8:23 am"
    # @option :none игнорирует значение времени
    # @param [Sym] god при русской локализации и методе :human добавляет "г." после даты
    # @return [String] строка с форматированными датой и временем
    # @example Примеры использования
    #   StTools.configure { |config| config.locale = :ru }
    #   StTools::Human.format_time2(Time.now, :human, :full)               #=> "30 апреля 2015 г. 08:54:34"
    #   StTools::Human.format_time2(Time.now, :human, :full, god: false)   #=> "30 апреля 2015 08:54:34"
    #   StTools::Human.format_time2(Time.now, :human, :short)              #=> "30 апреля 2015 г. 8:54"
    #   StTools::Human.format_time2(Time.now, :human, :none)               #=> "30 апреля 2015 г."
    #   StTools::Human.format_time2(Time.now, :full, :full)                #=> "30/04/2015 08:54:34"
    #   StTools::Human.format_time2(Time.now, :short, :short)              #=> "30/04/15 8:54"
    #   StTools::Human.format_time2(Time.now, :none, :full)                #=> "08:54:34"
    #   StTools::Human.format_time2(Time.now, :none, :short)               #=> "8:54"
    def self.format_time2(timestamp, date, time, god: true)
      unless [:human, :full, :short, :none].include?(date)
        warn "WARNING: date ':#{date.to_s}' must be in [:human, :full, :short, :none]. Use ':human' now (at line #{__LINE__} of StTools::#{File.basename(__FILE__)})"
        date = :human
      end
      unless [:full, :short, :none].include?(time)
        warn "WARNING: time ':#{time.to_s}' must be in [:full, :short, :none]. Use ':full' now (at line #{__LINE__} of StTools::#{File.basename(__FILE__)})"
        time = :full
      end

      case
        when date == :none && time == :none
          ""
        when date != :none && time == :none
          StTools::Human.format_time2_date(timestamp, date, god)
        when date != :none && time != :none
          "#{StTools::Human.format_time2_date(timestamp, date, god)} #{StTools::Human.format_time2_time(timestamp, time)}"
        when date == :none && time != :none
          StTools::Human.format_time2_time(timestamp, time)
        else
          ""
      end
    end

    # Метод оформляет число красивым способом, в виде "1 456 742,34".
    #
    # @param [Object] value исходное число в виде строки или числа, допустим nil
    # @param [Integer] round число цифр после запятой (по умолчанию - 0)
    # @param [Boolean] nobr обрамить результат тегами nobr (по умолчанию - false)
    # @param [Boolean] strong обрамить результат тегами strong (по умолчанию - false)
    # @return [String] строка с форматированным числом
    # @example Примеры использования
    #   StTools::Human.pretty_number(345)                       # => 345
    #   StTools::Human.pretty_number(345, round: 2)             # => 345,00
    #   StTools::Human.pretty_number(75345, round: 1)           # => 75 345,0
    #   StTools::Human.pretty_number(nil)                       # => 0
    #   StTools::Human.pretty_number('1675345.763', round: 1)   # => 1 675 345,7
    def self.pretty_number(value, round: 0, nobr: false, strong: false)
      out = StTools::String.to_float(value, round: round, stop: false).to_s
      arr = out.split(/[\,\.]/)
      tmp = arr.first.split(//).reverse.each_slice(3).to_a
      out = Array.new
      tmp.each do |one|
        out << one.reverse.join
      end
      out = out.reverse.join(' ')
      if arr.count > 1
        out = [out, (arr.last + '000000000000000')[0,round]].join(',')
      end
      out = "<nobr>#{out}</nobr>" if nobr
      out = "<strong>#{out}</strong>" if strong
      out
    end



    private




    def self.ago_in_words_pair(secs)
      mins = (secs / 60).to_i
      hours = (mins / 60).to_i
      days = (hours / 24).to_i
      years = (days / 365).to_i
      months = (days / (365/12)).to_i
      # puts "#{secs}, #{mins}, #{hours}, #{days}, #{months}, #{years}"

      return ago_in_words_one_pair(years, months - (years*12), "year", "month") if (months > 12)
      return ago_in_words_one_pair(months, days - (months*(365/12)), "month", "day") if (days > 28)
      return ago_in_words_one_pair(days, hours - (days*24), "day", "hour") if (hours > 24)
      return ago_in_words_one_pair(hours, mins - (hours*60), "hour", "minute") if (mins > 60)
      return ago_in_words_one_pair(mins, secs - (mins*60), "minute", "second") if (secs > 60)
      return [ago_in_words_one_value(secs, "second")]
    end

    def self.ago_in_words_one_pair(val1, val2, tag1, tag2)
      return [ago_in_words_one_value(val1, tag1), ago_in_words_one_value(val2, tag2)]
    end

    def self.ago_in_words_one_value(val, tag)
      num100 = val % 100
      num10 = val % 10
      return val.to_s + " " + I18n.t("common.ago.#{tag}.other", locale: StTools.configuration.locale) if (num100 >=5 && num100 <= 20)
      return val.to_s + " " + I18n.t("common.ago.#{tag}.one", locale: StTools.configuration.locale) if (num10 == 1)
      return val.to_s + " " + I18n.t("common.ago.#{tag}.two", locale: StTools.configuration.locale) if ([2, 3, 4].include?(num10))
      return val.to_s + " " + I18n.t("common.ago.#{tag}.other", locale: StTools.configuration.locale)
    end

    def self.to_time(time, form = :local)
      parts = Date._parse(time, false)
      return if parts.empty?

      now = Time.now
      time = Time.new(
          parts.fetch(:year, now.year),
          parts.fetch(:mon, now.month),
          parts.fetch(:mday, now.day),
          parts.fetch(:hour, 0),
          parts.fetch(:min, 0),
          parts.fetch(:sec, 0) + parts.fetch(:sec_fraction, 0),
          parts.fetch(:offset, form == :utc ? 0 : nil)
      )

      form == :utc ? time.utc : time.getlocal
    end

    def self.format_time2_time(timestamp, format)
      I18n.l(timestamp, :format => "t#{format.to_s}".to_sym, locale: StTools.configuration.locale)
    end

    def self.format_time2_date(timestamp, format, god)
      res = I18n.l(timestamp, :format => "d#{format.to_s}".to_sym, locale: StTools.configuration.locale)
      res += I18n.t("st_tools.god_socr", locale: StTools.configuration.locale) if god && format.to_sym == :human
      res
    end

  end
end
